File: app/src/main/java/ru/tubetv/app/VkVideoId.java
package ru.tubetv.app;

import android.net.Uri;

import org.json.JSONObject;

final class VkVideoId {
    private VkVideoId() { }

    static String fromUrl(String url) throws Exception {
        String id = findInText(url);
        Uri uri = Uri.parse(url);
        if (id != null) {
            if (hasAccessKey(id)) return id;
            String accessKey = queryAccessKey(uri);
            return accessKey.isEmpty() ? id : id + "_" + accessKey;
        }

        String owner = uri.getQueryParameter("oid");
        String video = uri.getQueryParameter("id");
        if (isSignedNumber(owner) && isUnsignedNumber(video)) {
            String accessKey = queryAccessKey(uri);
            return owner + "_" + video + (accessKey.isEmpty() ? "" : "_" + accessKey);
        }
        throw new Exception("Не найден ID VK Video");
    }

    static String fromVideo(JSONObject video) {
        String base = video.optLong("owner_id") + "_" + video.optLong("id");
        String accessKey = cleanAccessKey(video.optString("access_key"));
        if (accessKey.isEmpty()) accessKey = cleanAccessKey(video.optString("accessKey"));
        if (!accessKey.isEmpty()) return base + "_" + accessKey;

        String[] urlFields = {"direct_url", "share_url", "player"};
        for (String field : urlFields) {
            try {
                String candidate = fromUrl(video.optString(field));
                if (candidate.startsWith(base + "_")) return candidate;
            } catch (Exception ignored) { }
        }
        return base;
    }

    private static String queryAccessKey(Uri uri) {
        String accessKey = cleanAccessKey(uri.getQueryParameter("access_key"));
        return accessKey.isEmpty()
                ? cleanAccessKey(uri.getQueryParameter("hash")) : accessKey;
    }

    private static boolean hasAccessKey(String id) {
        int first = id.indexOf('_');
        return first >= 0 && id.indexOf('_', first + 1) >= 0;
    }

    private static String findInText(String value) {
        if (value == null || value.isEmpty()) return null;
        int from = 0;
        while (from < value.length()) {
            int marker = value.indexOf("video", from);
            if (marker < 0) return null;
            int start = marker + 5;
            int cursor = start;
            if (cursor < value.length() && value.charAt(cursor) == '-') cursor++;
            int ownerStart = cursor;
            while (cursor < value.length() && isDigit(value.charAt(cursor))) cursor++;
            if (cursor > ownerStart && cursor < value.length() && value.charAt(cursor) == '_') {
                cursor++;
                int videoStart = cursor;
                while (cursor < value.length() && isDigit(value.charAt(cursor))) cursor++;
                if (cursor > videoStart) {
                    if (cursor < value.length() && value.charAt(cursor) == '_') {
                        int accessStart = ++cursor;
                        while (cursor < value.length() && isAccessKeyChar(value.charAt(cursor))) cursor++;
                        if (cursor == accessStart) cursor--;
                    }
                    return value.substring(start, cursor);
                }
            }
            from = marker + 5;
        }
        return null;
    }

    private static String cleanAccessKey(String value) {
        if (value == null || value.isEmpty()) return "";
        int end = 0;
        while (end < value.length() && isAccessKeyChar(value.charAt(end))) end++;
        return end == 0 ? "" : value.substring(0, end);
    }

    private static boolean isSignedNumber(String value) {
        if (value == null || value.isEmpty()) return false;
        int start = value.charAt(0) == '-' ? 1 : 0;
        return start < value.length() && isDigits(value, start);
    }

    private static boolean isUnsignedNumber(String value) {
        return value != null && !value.isEmpty() && isDigits(value, 0);
    }

    private static boolean isDigits(String value, int start) {
        for (int i = start; i < value.length(); i++) if (!isDigit(value.charAt(i))) return false;
        return true;
    }

    private static boolean isDigit(char value) {
        return value >= '0' && value <= '9';
    }

    private static boolean isAccessKeyChar(char value) {
        return isDigit(value) || value >= 'a' && value <= 'z'
                || value >= 'A' && value <= 'Z' || value == '-' || value == '_';
    }
}

File: MacOS/Sources/App/Clients/VkVideoId.swift
import Foundation

/// Извлечение идентификатора `owner_id_video_id[_access_key]` из произвольных ссылок VK.
enum VkVideoId {
    static func fromUrl(_ url: String?) throws -> String {
        guard let url else { throw HttpError.message("Не найден ID VK Video") }
        if let id = findInText(url) {
            if hasAccessKey(id) { return id }
            let accessKey = queryAccessKey(url)
            return accessKey.isEmpty ? id : id + "_" + accessKey
        }
        let owner = Url.queryParameter(url, "oid")
        let video = Url.queryParameter(url, "id")
        if isSignedNumber(owner), isUnsignedNumber(video) {
            let accessKey = queryAccessKey(url)
            return owner! + "_" + video! + (accessKey.isEmpty ? "" : "_" + accessKey)
        }
        throw HttpError.message("Не найден ID VK Video")
    }

    static func fromVideo(_ video: JSON) -> String {
        let base = "\(video.long("owner_id"))_\(video.long("id"))"
        var accessKey = cleanAccessKey(video.string("access_key"))
        if accessKey.isEmpty { accessKey = cleanAccessKey(video.string("accessKey")) }
        if !accessKey.isEmpty { return base + "_" + accessKey }

        for field in ["direct_url", "share_url", "player"] {
            if let candidate = try? fromUrl(video.string(field)), candidate.hasPrefix(base + "_") {
                return candidate
            }
        }
        return base
    }

    private static func queryAccessKey(_ url: String) -> String {
        let accessKey = cleanAccessKey(Url.queryParameter(url, "access_key"))
        return accessKey.isEmpty ? cleanAccessKey(Url.queryParameter(url, "hash")) : accessKey
    }

    private static func hasAccessKey(_ id: String) -> Bool {
        guard let first = id.firstIndex(of: "_") else { return false }
        return id[id.index(after: first)...].contains("_")
    }

    private static func findInText(_ value: String) -> String? {
        let characters = Array(value)
        guard !characters.isEmpty else { return nil }
        var from = 0
        while from < characters.count {
            guard let marker = indexOf("video", in: characters, from: from) else { return nil }
            let start = marker + 5
            var cursor = start
            if cursor < characters.count && characters[cursor] == "-" { cursor += 1 }
            let ownerStart = cursor
            while cursor < characters.count && isDigit(characters[cursor]) { cursor += 1 }
            if cursor > ownerStart && cursor < characters.count && characters[cursor] == "_" {
                cursor += 1
                let videoStart = cursor
                while cursor < characters.count && isDigit(characters[cursor]) { cursor += 1 }
                if cursor > videoStart {
                    if cursor < characters.count && characters[cursor] == "_" {
                        cursor += 1
                        let accessStart = cursor
                        while cursor < characters.count && isAccessKeyChar(characters[cursor]) { cursor += 1 }
                        if cursor == accessStart { cursor -= 1 }
                    }
                    return String(characters[start..<cursor])
                }
            }
            from = marker + 5
        }
        return nil
    }

    private static func indexOf(_ needle: String, in characters: [Character], from: Int) -> Int? {
        let pattern = Array(needle)
        guard characters.count >= pattern.count else { return nil }
        var index = max(0, from)
        while index + pattern.count <= characters.count {
            if Array(characters[index..<(index + pattern.count)]) == pattern { return index }
            index += 1
        }
        return nil
    }

    private static func cleanAccessKey(_ value: String?) -> String {
        guard let value, !value.isEmpty else { return "" }
        let characters = Array(value)
        var end = 0
        while end < characters.count && isAccessKeyChar(characters[end]) { end += 1 }
        return end == 0 ? "" : String(characters[0..<end])
    }

    private static func isSignedNumber(_ value: String?) -> Bool {
        guard let value, !value.isEmpty else { return false }
        let body = value.hasPrefix("-") ? String(value.dropFirst()) : value
        return !body.isEmpty && body.allSatisfy(isDigit)
    }

    private static func isUnsignedNumber(_ value: String?) -> Bool {
        guard let value, !value.isEmpty else { return false }
        return value.allSatisfy(isDigit)
    }

    private static func isDigit(_ value: Character) -> Bool { value >= "0" && value <= "9" }

    private static func isAccessKeyChar(_ value: Character) -> Bool {
        isDigit(value) || (value >= "a" && value <= "z") || (value >= "A" && value <= "Z")
            || value == "-" || value == "_"
    }
}
